adminpage/sport/admin/trainingAdmin.py

```diff
@@ -169,14 +178,21 @@ def get_fieldsets(self, request, obj=None):
 
     @transaction.atomic
     def save_related(self, request, form, formsets, change):
+        """
+        There are three ways of creating/changing attendance records (from highest priority to lowest):
+        1) Upload CSV file
+        2) Use ModelMultipleChoiceField in ModelAdmin for training
+        3) Change/add attendance records using inline
+        """
         super().save_related(request, form, formsets, change)
         training = form.instance
-        for student in form.cleaned_data['attended_students']:
-            Attendance.objects.update_or_create(student=student, training=training,
-                                                defaults={'hours': form.cleaned_data['hours']})
-        for (student, hours) in form.cleaned_data['attendances']:
-            if hours == 0:
-                Attendance.objects.filter(student=student, training=training).delete()
-            else:
-                Attendance.objects.update_or_create(student=student, training=training, defaults={'hours': hours})
+        similar_student_hours = [
+            (student.pk, form.cleaned_data['hours'])
+            for student in form.cleaned_data['attended_students']
+        ]
+        csv_student_hours = [
+            (student.pk, hours)
+            for (student, hours) in form.cleaned_data['attendances']
+        ]
+        mark_hours(training, similar_student_hours + csv_student_hours)
```

Comment:
if the same student is present in both: `similar_student_hours` and `csv_student_hours` will this case be handled by postgres correctly?

Reply:
I fixed this. Now only one row from `csv_student_hours` should be kept